6 migration stage 2 migrate transform module

suarezgpablo opened the pull request:
Added Java files and test files from the tranform module of Gitlab to Github

suarezgpablo commented:
@javiertuya I migrated the files and the tests seem to be working. However, Sonarcloud is giving false positives for private fields where lombok is used for the Setters and Getters. I searched on the Internet and it seems to be a known issue that has not been solved yet:

https://sonarsource.atlassian.net/browse/SONARJAVA-4178

javiertuya commented:
@suarezgpablo Maybe you have removed from sonar-project.properties the line that avoids this problem?. Compare this file with the one that is in GitLab

suarezgpablo commented:
> @suarezgpablo Maybe you have removed from sonar-project.properties the line that avoids this problem?. Compare this file with the one that is in GitLab

Yes, it was that, thanks. I think that I removed the property when customizing the template because I thought that was causing some issues and didn't remember to put it back when everything else was fixed.

javiertuya commented on pom.xml:
```diff
@@ -46,6 +46,12 @@
+				<groupId>org.projectlombok</groupId>
+				<artifactId>lombok</artifactId>
+				<version>1.18.32</version>
+				<scope>provided</scope>
```
Why scope provided?

suarezgpablo replied:
That's how I saw it recommeded in the lombok webpage. I read that there is also the "compile" scope, would "compile" be better?

suarezgpablo replied:
Do I keep it with the scope provided?

javiertuya replied:
ok, keep provided, if you later need lombok in the script module, just declare it.

javiertuya commented on pom.xml:
```diff
@@ -109,6 +115,45 @@
 						</execution>
 					</executions>
 				</plugin>
+				<plugin>
+				<groupId>org.jvnet.maven-antrun-extended-plugin</groupId>
+				<artifactId>maven-antrun-extended-plugin</artifactId>
+				<version>1.43</version>
+				<executions>
```
This plugin seems to be duplicating the report genertion, see the next plugin

suarezgpablo replied:
Was this about the issue of the double compilation?

javiertuya replied:
The issue was that the large number of warnings make difficult to know what maven is doing if three are two places where the warnings are shown. Now that warnings are removed it is not a major problem. I'm inclined to remove the aggregate reporting for this project, but it can be postponed to the migration is completed. There is a 4th stage to check  and reorganize the reports. Go ahead as is.

javiertuya commented:
@suarezgpablo Please, check the log of the test job, I see two issues:
- ATL compilation is giving lots of warnings
- ATL compilation is executing twice

suarezgpablo commented:
> @suarezgpablo Please, check the log of the test job, I see two issues:
> 
>     * ATL compilation is giving lots of warnings
> 
>     * ATL compilation is executing twice
> 
> 
> Alternatively, create an issue to solve after this PR is merged

Yes, I'll look into the warnings to see how to remove them.

The ATL compilation on the other hand I only see that it is executed once (from line 173 to 302, being most of them the aforementioned warnings), but perhaps I'm missing something.

suarezgpablo commented:
> > @suarezgpablo Please, check the log of the test job, I see two issues:
> > ```
> > * ATL compilation is giving lots of warnings
> > 
> > * ATL compilation is executing twice
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Alternatively, create an issue to solve after this PR is merged
> 
> Yes, I'll look into the warnings to see how to remove them.

I saw now the double compilation of ATL in test-java, I was looking only in Sonarqube

suarezgpablo commented:
> > > @suarezgpablo Please, check the log of the test job, I see two issues:
> > > ```
> > > * ATL compilation is giving lots of warnings
> > > 
> > > * ATL compilation is executing twice
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Alternatively, create an issue to solve after this PR is merged
> > 
> > 
> > Yes, I'll look into the warnings to see how to remove them.
> 
> I saw now the double compilation of ATL in test-java, I was looking only in Sonarqube. Actually I think that it is executing everything twice, not just the ATL compilation. I'll look into that too.

I think the problem is because when it executes the pom of the script module it executes again the plugins specified in the pom of the transform module, I assume because of the dependency.

javiertuya commented:
> I think the problem is because when it executes the pom of the script module it executes again the plugins specified in the pom of the transform module, I assume because of the dependency.

I don't think so, because when executing the script module:
```
[INFO] ----------------< io.github.giis-uniovi:modevo-script >-----------------
[INFO] Building modevo-script 1.0.1-SNAPSHOT                              [3/4]
[INFO]   from modevo-script/pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
```

it has already compiled the ATL twice

suarezgpablo commented:
> > I think the problem is because when it executes the pom of the script module it executes again the plugins specified in the pom of the transform module, I assume because of the dependency.
> 
> I don't think so, because when executing the script module:
> 
> ```
> [INFO] ----------------< io.github.giis-uniovi:modevo-script >-----------------
> [INFO] Building modevo-script 1.0.1-SNAPSHOT                              [3/4]
> [INFO]   from modevo-script/pom.xml
> [INFO] --------------------------------[ jar ]---------------------------------
> ```
> 
> it has already compiled the ATL twice

Yes, you were right. I removed the "surefire-report:report" goal in the maven run locally and then it executed it only once. I'll look into it

suarezgpablo commented:
@javiertuya Regarding the double execution of the plugins, I looked in the actions of tdrules and I think there is the same behaviour there. After the line that contains "maven-surefire-report-plugin:3.2.5:report-only (ut-reports) @ tdrules-model/modevo-transform" the plugins are compiled again both in modevo and in tdrules, but not the tests. Could there be a requirement of this goal to compile again everything?

suarezgpablo commented:
@javiertuya I fixed the warnings issued by the Atl compiler. I added a Maven job to create a jar file from the java class that I was using:

`<plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-jar-plugin</artifactId>
                <version>3.2.0</version>
                <executions>
                    <execution>
                        <id>make-jar</id>
                        <phase>compile</phase>
                        <goals>
                            <goal>jar</goal>
                        </goals>
		                <configuration>
							<finalName>AtlCompilerTask</finalName>
		                    <includes>
		                        <include>${basedir}/src/main/java/giis/modevo/transformations/atl/AtlCompilerTask.java</include>
		                    </includes>
		                    <archive>
		                        <manifest>
		                            <mainClass>giis.modevo.transformations.atl.AtlCompilerTask</mainClass>
		                        </manifest>
		                    </archive>
		                </configuration>
		          </execution>
                </executions>         
            </plugin>`

With this, the jar for the compilation of ATL files is created in the folder target. Let me know if this needs to be changed

javiertuya commented:
@suarezgpablo I have checked it, yo are right. The duplication is produced by the duble target mvn test report: `mvn test surefire-report:report -Daggregate=true`. What happens is that in modevo is more noticeable because of the warnings. If I remember well, in tdrules, it is needed to consolidate the reports from all modules.

In the case of modevo, I think that you can remove this and keep `mvn test` only, because the script module will run in a different job and the aggregation of reports will be done differently.

javiertuya commented on modevo-transform/pom.xml:
```diff
@@ -87,6 +87,31 @@
+                <version>3.2.0</version>
+                <executions>
+                    <execution>
+                        <id>make-jar</id>
+                        <phase>compile</phase>
```
I think that this is better in a previous phase, eg. `process-resources`. Put a more significant name, as this is not the module jar, but something that is necessary for compilation.

suarezgpablo replied:
@javiertuya I changed it to process-resources and it runs ok in my computer using the same goals as in GitHub. However, in GitHub it is not able to create the Jar file:

`Warning:  JAR will be empty - no content was marked for inclusion!`

The run was the following one:

https://github.com/giis-uniovi/modevo/actions/runs/8572462627

I returned it back to the compile phase for now, which seems to be running ok. Do I keep looking into it or de we keep it in compile?

javiertuya replied:
process-resources is the latest phase before compile, so, keep it as is
